EventSource/src/EventSourceUtil.ts

```diff
@@ -58,6 +55,150 @@ export function getTimeStamp(resource: Resource, timestampPath: string): number
+                    // the IDs (which would in turn not be memory efficient)
+                    if (obj.equals(quad.object)) {
+                        continue addElements;
+                    }
+                }
+                objs.push(quad.object);
+            } else {
+                props.set(quad.predicate.id, new Array(quad.object));
+            }
+        } else {
+            data.set(quad.subject.id, new Map([[quad.predicate.id, new Array(quad.object)]]));
+        }
+    }
+    return [named, blank];
+}
+
+const invalidPrefixCharacters = ["/", ":"];
+
+/**
+ * Converts a quad-subj/pred/obj ID to a potentially shorter version by using a compatible prefix;
+ * otherwise a turtle friendly version is returned
+ * @param original
+ * @returns {string}
+ */
+export function tryApplyPrefix(original: string, prefixes: any): string {
+    // the string might either be a direct value (e.g. saref:hasValue)
+    // or represent a literal (e.g. "0.0"^^xsd:float), can be checked
+    // by looking for regex
+    if (/\"[^"]*\"\^\^[^\^]+/.test(original)) {
+        // only checking after the ^^ for prefixes
+        const str = original.substring(original.indexOf("^^") + 2);
+        for (const prefix in prefixes) {
+            if (str.startsWith(prefixes[prefix])) {
+                const substr = str.substring(prefixes[prefix].length);
+                if (!invalidPrefixCharacters.some(char => substr.includes(char))) {
+                    return `${original.substring(0, original.indexOf("^^") + 2)}${prefix}:${substr}`;
+                }
+            }
+        }
+        return `${original.substring(0, original.indexOf("^^") + 2)}<${str}>`;
+    } else {
+        for (const prefix in prefixes) {
+            if (original.startsWith(prefixes[prefix])) {
+                const substr = original.substring(prefixes[prefix].length);
+                if (!invalidPrefixCharacters.some(char => substr.includes(char))) {
+                    return `${prefix}:${substr}`;
+                }
+            }
+        }
+        return `<${original}>`;
+    }
+}
+
+/**
+ * Converts a resource (quad array) to an optimised turtle string representation by grouping subjects
+ * together, using prefixes wherever possible and replacing blank nodes with their properties.
+ * Note: blank nodes referenced as objects, but not found as subjects in other quads, are removed
+ *  entirely
+ *
+ * @param resource The resource that gets converted to a string
+ * @param prefixes An object which members are strings, member name being the short prefix and its
+ *  value a string representing its URI. Example: `{"rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#"}`
+ * @param skipPrefixDecl If no \@prefix x: <y> . on the top is required, this can be set to true
+ * @returns {string}
+ */
+export function resourceToOptimisedTurtle(resource: Resource, prefixes: any, skipPrefixDecl: boolean = false): string {
```

Comment:
It is better to use the [Writer](https://github.com/rdfjs/N3.js#writing) class from N3.js. Just like [this](https://github.com/woutslabbinck/VersionAwareLDESinLDP/blob/main/src/util/Conversion.ts#L31-L32) utility does as a basis.

As you can in the examples in Writer, it is possible to just pass the prefixes and it does the job for you (as a bonus, this method was also been tested).